fabric/patches/2_0001-Extension-point-for-additional-SCCs-opt_scc.patch

```diff
@@ -0,0 +1,172 @@
++
++func (s *mock_scc) Name() string {
++	return s.name
++}
++func (s *mock_scc) Chaincode() shim.Chaincode {
++	return nil
++}
++
++func TestRegistry(t *testing.T) {
++	g := NewGomegaWithT(t)
++
++	// 1. initially we should get an empty list
++	ffs := opt_scc.ListFactoryFuncs()
++	g.Expect(ffs).NotTo(BeNil(), "factory function list should not be nil")
++	g.Expect(len(ffs)).To(Equal(0), "factory function list should have zero elements")
++
++	// 2. after adding several handlers, make sure it returns exactly and only these handlers
++	names := []string{"ff1", "ff2", "ff3"}
++	for _, n := range names {
++		nLocal := n // to get separate version for below closure ...
++		opt_scc.AddFactoryFunc(
++			func(aclProvider aclmgmt.ACLProvider, p *peer.Peer) scc.SelfDescribingSysCC {
++				return &mock_scc{nLocal}
++			})
++	}
++	ffs = opt_scc.ListFactoryFuncs()
++	g.Expect(ffs).NotTo(BeNil(), "factory function list should not be nil")
++	g.Expect(len(ffs)).To(Equal(len(names)), "factory function list should have %v elements", len(names))
++	got_names := []string{}
++	for i, ff := range ffs {
++		s := ff(nil, nil)
++		g.Expect(s).NotTo(BeNil(), "received unexpected factory function (element %v in list) returns nil", i)
++		got_names = append(got_names, s.Name())
++	}
++	// Note: we do not force any order, so compare sorted lists
++	sort.Strings(names)
++	sort.Strings(got_names)
++	g.Expect(got_names).To(Equal(names), "Unexpected factory function in list")
++
++	// 3. Test that error on nil factory functions
++	g.Expect(func() {
++		opt_scc.AddFactoryFunc(nil)
++	}).To(Panic(), "Adding a nil factory function should panic")
++
++}
+diff --git a/internal/peer/node/start.go b/internal/peer/node/start.go
+index 63c0ae919..5217ce540 100644
+--- a/internal/peer/node/start.go
++++ b/internal/peer/node/start.go
+@@ -73,6 +73,7 @@ import (
+ 	"github.com/hyperledger/fabric/core/scc"
+ 	"github.com/hyperledger/fabric/core/scc/cscc"
+ 	"github.com/hyperledger/fabric/core/scc/lscc"
++	"github.com/hyperledger/fabric/core/scc/opt_scc"
+ 	"github.com/hyperledger/fabric/core/scc/qscc"
+ 	"github.com/hyperledger/fabric/core/transientstore"
+ 	"github.com/hyperledger/fabric/discovery"
+@@ -595,6 +596,16 @@ func serve(args []string) error {
+ 		"_lifecycle": {},
+ 	}
+ 
++	optSCCInsts := []scc.SelfDescribingSysCC{}
++	for _, ff := range opt_scc.ListFactoryFuncs() {
++		osi := ff(aclProvider, peerInstance)
++		if osi == nil {
++			logger.Panic("Optional SCC factory function returning nil SCC")
++		}
++		optSCCInsts = append(optSCCInsts, osi)
++		builtinSCCs[osi.Name()] = struct{}{}
++	}
++
+ 	lsccInst := &lscc.SCC{
+ 		BuiltinSCCs: builtinSCCs,
+ 		Support: &lscc.SupportImpl{
+@@ -744,7 +755,7 @@ func serve(args []string) error {
+ 	}
+ 
+ 	// deploy system chaincodes
+-	for _, cc := range []scc.SelfDescribingSysCC{lsccInst, csccInst, qsccInst, lifecycleSCC} {
++	for _, cc := range append([]scc.SelfDescribingSysCC{lsccInst, csccInst, qsccInst, lifecycleSCC}, optSCCInsts...) {
```

Comment:
I wonder if this could be better factored by creating a slice of "requiredSCCInsts" to act as a counter to "optionalSCCInsts".

Not a fan of "opt" and "insts" but that is your code style.

Reply:
The `insts` come from original fabric code in `internal/node/peer/start.go`.  The `opt` i'm also not super enamored with but couldn't come up with something really better ...